src/execution/BucketExecutor.tsx

```diff
@@ -0,0 +1,338 @@
+    limit: number[];
+}
+
+export interface IBucketExecutorChildrenProps {
+    isLoading: boolean;
+    error: any;
+    result: Execution.IExecutionResult | null;
+    response: Execution.IExecutionResponse | null;
+    getPage: (paging: IPaging) => void;
+}
+
+export interface IBucketExecutorProps {
+    telemetryComponentName?: string;
+    autoLoadFirstPage?: boolean;
+    initialPaging?: IPaging;
+    sdk?: SDK;
+
+    children: (childrenProps: IBucketExecutorChildrenProps) => React.ReactNode;
+    dimensions: dimensionItem[][];
+    projectId: string;
+    filters?: VisualizationInput.IFilter[];
+    sortBy?: VisualizationInput.ISort[];
+}
+
+export interface IBucketExecutorState {
+    result: Execution.IExecutionResult | null;
+    response: Execution.IExecutionResponse | null;
+    isLoading: boolean;
+    error: any;
+    responsePromise: Promise<Execution.IExecutionResponse> | null;
+    resultPromise: Promise<Execution.IExecutionResult> | null;
+
+    children: IBucketExecutorProps["children"];
+    dimensions: IBucketExecutorProps["dimensions"];
+    projectId: IBucketExecutorProps["projectId"];
+    filters: IBucketExecutorProps["filters"];
+    sortBy: IBucketExecutorProps["sortBy"];
+}
+
+export interface IBucketExecutorChildrenProps {
+    response: Execution.IExecutionResponse | null;
+    result: Execution.IExecutionResult | null;
+    isLoading: boolean;
+    error: any;
+    getPage: (paging: IPaging) => void;
+}
+
+const propsRequiringDataReset: Array<keyof IBucketExecutorProps> = [
+    "projectId",
+    "dimensions",
+    "filters",
+    "sortBy",
+];
+
+const didPropsChange = <V extends {}>(propList: string[], prevProps: V, props: V): boolean => {
+    for (const propName of propList) {
+        const prevProp = prevProps[propName];
+        const prop = props[propName];
+        if (prop !== prevProp) {
+            return true;
+        }
+    }
+    return false;
+};
+
+const setupSDK = (props: IBucketExecutorProps) => {
+    const clonedSdk = props.sdk ? props.sdk.clone() : createSdk();
+    setTelemetryHeaders(clonedSdk, props.telemetryComponentName, props);
+    return clonedSdk;
+};
+
+export class BucketExecutor extends React.Component<IBucketExecutorProps, IBucketExecutorState> {
+    public static defaultProps: Partial<IBucketExecutorProps> = {
+        sdk: createSdk(),
+        filters: [],
+        sortBy: [],
+        autoLoadFirstPage: true,
+        telemetryComponentName: "BucketExecutor",
+    };
+
+    public static getDerivedStateFromProps(nextProps: IBucketExecutorProps, prevState: IBucketExecutorState) {
+        const newState: Partial<IBucketExecutorState> = {};
+        const propsToUpdate = [...propsRequiringDataReset, "children"];
+        propsToUpdate.forEach(propName => {
+            if (nextProps[propName] !== prevState[propName]) {
+                newState[propName] = nextProps[propName];
+            }
+        });
+        if (Object.keys(newState).length === 0) {
```

Comment:
Nitpick: you could keep a track of a flag in the above `forEach` (something like `hasPropBeenSynced`). Then you wouldn't need this call.